File: python-service/src/calculator_service.py
"""
Calculator Service for pytest unit testing
"""

import math
from typing import Union, List


class CalculatorService:
    """Service class for mathematical calculations"""
    
    def add(self, a: Union[int, float], b: Union[int, float]) -> float:
        """Add two numbers"""
        return float(a + b)
    
    def subtract(self, a: Union[int, float], b: Union[int, float]) -> float:
        """Subtract two numbers"""
        return float(a - b)
    
    def multiply(self, a: Union[int, float], b: Union[int, float]) -> float:
        """Multiply two numbers"""
        return float(a * b)
    
    def divide(self, a: Union[int, float], b: Union[int, float]) -> float:
        """Divide two numbers"""
        if b == 0:
            raise ValueError("Division by zero is not allowed")
        return float(a / b)
    
    def power(self, base: Union[int, float], exponent: Union[int, float]) -> float:
        """Calculate power of a number"""
        return float(base ** exponent)
    
    def sqrt(self, number: Union[int, float]) -> float:
        """Calculate square root of a number"""
        if number < 0:
            raise ValueError("Cannot calculate square root of negative number")
        return float(math.sqrt(number))
    
    def factorial(self, n: int) -> int:
        """Calculate factorial of a number"""
        if n < 0:
            raise ValueError("Factorial is not defined for negative numbers")
        if n == 0 or n == 1:
            return 1
        return n * self.factorial(n - 1)
    
    def average(self, numbers: List[Union[int, float]]) -> float:
        """Calculate average of a list of numbers"""
        if not numbers:
            raise ValueError("Cannot calculate average of empty list")
        return float(sum(numbers) / len(numbers))
    
    def max_number(self, numbers: List[Union[int, float]]) -> Union[int, float]:
        """Find maximum number in a list"""
        if not numbers:
            raise ValueError("Cannot find maximum of empty list")
        return max(numbers)
    
    def min_number(self, numbers: List[Union[int, float]]) -> Union[int, float]:
        """Find minimum number in a list"""
        if not numbers:
            raise ValueError("Cannot find minimum of empty list")
        return min(numbers)
    
    def gcd(self, a: int, b: int) -> int:
        """Calculate greatest common divisor"""
        if a == 0:
            return abs(b)
        if b == 0:
            return abs(a)
        return self.gcd(b, a % b)
    
    def lcm(self, a: int, b: int) -> int:
        """Calculate least common multiple"""
        if a == 0 or b == 0:
            return 0
        return abs(a * b) // self.gcd(a, b)
    
    def is_prime(self, n: int) -> bool:
        """Check if a number is prime"""
        if n < 2:
            return False
        if n == 2:
            return True
        if n % 2 == 0:
            return False
        for i in range(3, int(math.sqrt(n)) + 1, 2):
            if n % i == 0:
                return False
        return True

File: node-service/src/calculatorService.js
/**
 * Calculator Service for Jest unit testing
 */

class CalculatorService {
  /**
   * Add two numbers
   * @param {number} a - First number
   * @param {number} b - Second number
   * @returns {number} Sum of the two numbers
   */
  add(a, b) {
    return a + b;
  }

  /**
   * Subtract two numbers
   * @param {number} a - First number
   * @param {number} b - Second number
   * @returns {number} Difference of the two numbers
   */
  subtract(a, b) {
    return a - b;
  }

  /**
   * Multiply two numbers
   * @param {number} a - First number
   * @param {number} b - Second number
   * @returns {number} Product of the two numbers
   */
  multiply(a, b) {
    return a * b;
  }

  /**
   * Divide two numbers
   * @param {number} a - First number
   * @param {number} b - Second number
   * @returns {number} Quotient of the two numbers
   * @throws {Error} When dividing by zero
   */
  divide(a, b) {
    if (b === 0) {
      throw new Error('Division by zero is not allowed');
    }
    return a / b;
  }

  /**
   * Calculate power of a number
   * @param {number} base - Base number
   * @param {number} exponent - Exponent
   * @returns {number} Result of base raised to exponent
   */
  power(base, exponent) {
    return Math.pow(base, exponent);
  }

  /**
   * Calculate square root of a number
   * @param {number} number - Number to find square root of
   * @returns {number} Square root of the number
   * @throws {Error} When number is negative
   */
  sqrt(number) {
    if (number < 0) {
      throw new Error('Cannot calculate square root of negative number');
    }
    return Math.sqrt(number);
  }

  /**
   * Calculate factorial of a number
   * @param {number} n - Number to calculate factorial of
   * @returns {number} Factorial of the number
   * @throws {Error} When number is negative
   */
  factorial(n) {
    if (n < 0) {
      throw new Error('Factorial is not defined for negative numbers');
    }
    if (n === 0 || n === 1) {
      return 1;
    }
    return n * this.factorial(n - 1);
  }

  /**
   * Calculate average of an array of numbers
   * @param {number[]} numbers - Array of numbers
   * @returns {number} Average of the numbers
   * @throws {Error} When array is empty
   */
  average(numbers) {
    if (!numbers || numbers.length === 0) {
      throw new Error('Cannot calculate average of empty array');
    }
    const sum = numbers.reduce((acc, num) => acc + num, 0);
    return sum / numbers.length;
  }

  /**
   * Find maximum number in an array
   * @param {number[]} numbers - Array of numbers
   * @returns {number} Maximum number
   * @throws {Error} When array is empty
   */
  max(numbers) {
    if (!numbers || numbers.length === 0) {
      throw new Error('Cannot find maximum of empty array');
    }
    return Math.max(...numbers);
  }

  /**
   * Find minimum number in an array
   * @param {number[]} numbers - Array of numbers
   * @returns {number} Minimum number
   * @throws {Error} When array is empty
   */
  min(numbers) {
    if (!numbers || numbers.length === 0) {
      throw new Error('Cannot find minimum of empty array');
    }
    return Math.min(...numbers);
  }

  /**
   * Calculate greatest common divisor
   * @param {number} a - First number
   * @param {number} b - Second number
   * @returns {number} Greatest common divisor
   */
  gcd(a, b) {
    if (a === 0) return Math.abs(b);
    if (b === 0) return Math.abs(a);
    return this.gcd(b, a % b);
  }

  /**
   * Calculate least common multiple
   * @param {number} a - First number
   * @param {number} b - Second number
   * @returns {number} Least common multiple
   */
  lcm(a, b) {
    if (a === 0 || b === 0) return 0;
    return Math.abs(a * b) / this.gcd(a, b);
  }

  /**
   * Check if a number is prime
   * @param {number} n - Number to check
   * @returns {boolean} True if prime, false otherwise
   */
  isPrime(n) {
    if (n < 2) return false;
    if (n === 2) return true;
    if (n % 2 === 0) return false;
    
    for (let i = 3; i <= Math.sqrt(n); i += 2) {
      if (n % i === 0) return false;
    }
    return true;
  }
}

module.exports = CalculatorService;
